src/algebraic_geometry/elliptic_curve/weierstrass.lean

```diff
@@ -381,19 +385,62 @@ https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/.E2.9C.94.20c
 instance [is_domain R] [normalized_gcd_monoid R] : is_domain W.coordinate_ring :=
 (ideal.quotient.is_domain_iff_prime _).mpr $
 by simpa only [ideal.span_singleton_prime W.polynomial_ne_zero, ← gcd_monoid.irreducible_iff_prime]
-   using W.polynomial_irreducible
-
-instance coordinate_ring.is_domain_of_field {F : Type u} [field F] (W : weierstrass_curve F) :
-  is_domain W.coordinate_ring :=
-by { classical, apply_instance }
+   using W.irreducible_polynomial
 
 /-- The function field $R(W) := \mathrm{Frac}(R[W])$ of `W`. -/
 @[reducible] def function_field : Type u := fraction_ring W.coordinate_ring
 
+variables (x : R) (y : R[X])
+
+/-- The class of the element $X - x$ in $R[W]$ for some $x \in R$. -/
+@[simp] noncomputable def X_class : W.coordinate_ring := adjoin_root.mk W.polynomial $ C $ X - C x
+
+/-- The class of the element $Y - y(X)$ in $R[W]$ for some $y(X) \in R[X]$. -/
+@[simp] noncomputable def Y_class : W.coordinate_ring := adjoin_root.mk W.polynomial $ X - C y
+
+/-- The ideal $\langle X - x \rangle$ of $R[W]$ for some $x \in R$. -/
+@[simp] noncomputable def X_ideal : ideal W.coordinate_ring := ideal.span {W.X_class x}
+
+/-- The ideal $\langle Y - y(X) \rangle$ of $R[W]$ for some $y(X) \in R[X]$. -/
+@[simp] noncomputable def Y_ideal : ideal W.coordinate_ring := ideal.span {W.Y_class y}
+
 end polynomial
 
 end weierstrass_curve
 
+namespace weierstrass_curve
+
+open polynomial
+
+open_locale non_zero_divisors polynomial
+
+variables {F : Type u} [field F] (W : weierstrass_curve F) (x : F) (y : F[X])
+
+instance coordinate_ring.is_domain_of_field : is_domain W.coordinate_ring :=
+by { classical, apply_instance }
+
+lemma X_class_ne_zero : W.X_class x ≠ 0 :=
+begin
+  intro hx,
+  cases ideal.mem_span_singleton'.mp (ideal.quotient.eq_zero_iff_mem.mp hx) with _ hx,
+  apply_fun degree at hx,
+  rw [degree_mul, degree_polynomial, degree_C $ X_sub_C_ne_zero x] at hx,
+  exact two_ne_zero (nat.with_bot.add_eq_zero_iff.mp hx).right
+end
+
+lemma Y_class_ne_zero : W.Y_class y ≠ 0 :=
+begin
+  intro hy,
+  cases ideal.mem_span_singleton'.mp (ideal.quotient.eq_zero_iff_mem.mp hy) with _ hy,
+  apply_fun degree at hy,
+  rw [degree_mul, degree_polynomial, degree_X_sub_C, nat.with_bot.add_eq_one_iff] at hy,
+  cases hy with hy hy,
+  { exact @order.succ_ne_succ (with_bot ℕ) _ _ _ _ _ one_ne_zero hy.right },
+  { exact two_ne_zero hy.right }
+end
```

Comment:
The following proofs fit in the previous `comm_ring` section. Seems there's no need to introduce a `field` section for this PR, but you may keep it if you need it in other PRs.
```suggestion
lemma X_class_ne_zero [nontrivial R] : W.X_class x ≠ 0 :=
begin
  intro hx,
  cases ideal.mem_span_singleton'.mp (ideal.quotient.eq_zero_iff_mem.mp hx) with _ hx',
  apply_fun degree at hx',
  rw [W.monic_polynomial.degree_mul, degree_polynomial, degree_C $ X_sub_C_ne_zero x] at hx',
  cases w.degree; cases hx',
end

lemma Y_class_ne_zero [nontrivial R] : W.Y_class y ≠ 0 :=
begin
  intro hy,
  cases ideal.mem_span_singleton'.mp (ideal.quotient.eq_zero_iff_mem.mp hy) with _ hy',
  apply_fun degree at hy',
  rw [W.monic_polynomial.degree_mul, W.degree_polynomial, degree_X_sub_C] at hy',
  cases w.degree; cases hy',
end
```

Reply:
I don't understand the magic at the end but thanks!

Reply:
It's abuse of defeq to some degree, but convenient. Same magic as 
```
lemma foo (h : 0 = 1) : false := by cases h
#print foo
/- theorem foo : 0 = 1 → false :=
λ (h : 0 = 1), h.dcases_on (λ (a : 1 = 0), nat.no_confusion a) (eq.refl 1) (heq.refl h) -/
```
Here `w.degree : with_bot nat` has two constructors so we need an additional `cases`. We're lucky here because `n + 2` unfolds to constructors `n.succ.succ`. If it's `2 + n` then it wouldn't work:
```
lemma foo (n : ℕ) (h : n + 2 = 1) : false := by cases h --works
lemma foo (n : ℕ) (h : 2 + n = 1) : false := by cases h
```